### DISCUSSION TITLE
module: How to express build time inter-module dependencies

### DISCUSSION THREAD

**Discussion by yashi** [zephyrproject-rtos/zephyr#50410] [Q&A] [answered]
I have two libraries, modA and modB.  modA depends on modB, that is, modA includes modB's header file, `include/modb.h`.  The header file is generated at build time of modB.

[`depends:`](https://docs.zephyrproject.org/latest/develop/modules.html#zephyr-module-dependencies) in `modA/zephyr/module.yml` _seems_ to work with `west build ... -o='-j1'` but doesn't seem to manage the order of builds when parallel.  Is there a way to express build time dependencies between two Zephyr modules?

- Zephyr: v3.2.0-rc2-69-g32f4f8d987f
- SDK: 0.15.0
- West: v0.14.0
- CMake: 3.24.2-1
- Ninja: 1.11.1
- OS: Debian Sid

**Comment by henrikbrixandersen** [comment]
Please see https://docs.zephyrproject.org/latest/develop/modules.html#zephyr-module-dependencies. I am not sure why that shouldn’t be sufficient for your use-case.

**Comment by tejlmand** [comment, marked as answer]
> but doesn't seem to manage the order of builds when parallel. Is there a way to express build time dependencies between two Zephyr modules?

The `depends: <module>` in module.yml is only ensuring:
- The referred module is available in the system
- The CMakeLists.txt file in the referred module will be processed before the CMakeLists.txt file in this module.
- Additional roots in the referred module will take precedence over roots in this module

there are no build time dependencies between CMake targets created in the two modules.

Target dependencies to ensure correct build order must be controlled using CMake dependency handling, for example:
https://cmake.org/cmake/help/latest/command/add_dependencies.html
